rust/otap-dataflow/crates/otap/src/tls_utils.rs

```diff
@@ -0,0 +1,696 @@
+///
+/// This function handles the TLS handshake for each incoming connection.
+/// TLS handshake failures are logged and filtered out (non-fatal).
+/// Transport-level listener errors are propagated to terminate the server.
+pub fn create_tls_stream<S, T>(
+    listener_stream: S,
+    tls_acceptor: tokio_rustls::TlsAcceptor,
+) -> impl Stream<Item = Result<tokio_rustls::server::TlsStream<T>, io::Error>>
+where
+    S: Stream<Item = Result<T, io::Error>> + Send + 'static,
+    T: tokio::io::AsyncRead + tokio::io::AsyncWrite + Send + Unpin + 'static,
+{
+    listener_stream.filter_map(move |conn_res| {
+        let acceptor = tls_acceptor.clone();
+        async move {
+            match conn_res {
+                Ok(conn) => {
+                    // Try TLS handshake
+                    match acceptor.accept(conn).await {
+                        Ok(stream) => Some(Ok::<_, io::Error>(stream)),
+                        Err(e) => {
+                            // TLS handshake failed - log and continue
+                            log::warn!("TLS handshake failed: {}", e);
+                            None
+                        }
+                    }
+                }
+                Err(e) => {
+                    // Transport-level listener error - propagate to terminate server
+                    Some(Err(e))
+                }
+            }
+        }
+    })
+}
+
+/// A certificate resolver that lazily reloads TLS certificates with throttled file modification time (mtime) checks.
+///
+/// # Performance characteristics
+///
+/// - Mtime checks for the certificate and key files are throttled by a configurable interval (`check_interval_secs`)
+///   to minimize filesystem operations. This means the filesystem is not polled on every handshake, but only after
+///   the specified interval has elapsed since the last check.
+/// - If a file change is detected during a TLS handshake (i.e., the mtime has changed since the last check),
+///   the certificate and key are reloaded asynchronously. This ensures that new certificates are picked up without
+///   requiring a server restart, while avoiding excessive filesystem access.
+/// - Reloads happen in the context of TLS handshakes. Certificate updates are applied to subsequent connections after
+///   the async reload completes, which typically happens within a short time after file modification is detected.
+#[derive(Debug)]
+pub struct LazyReloadableCertResolver {
+    /// Current certificate
+    cert_key: Arc<ArcSwap<CertifiedKey>>,
+
+    /// File paths
+    cert_path: PathBuf,
+    key_path: PathBuf,
+
+    /// Last known modification times (Arc for sharing with async reload tasks)
+    cert_mtime: Arc<AtomicU64>,
+    key_mtime: Arc<AtomicU64>,
+
+    /// Last time we checked mtime (unix timestamp in seconds)
+    last_check_time: AtomicU64,
+
+    /// Minimum interval between mtime checks (seconds)
+    check_interval_secs: u64,
```

Comment:
I would prefer std::time::Duration to be consistent across the code.

Reply:
I have updated the public API to use `std::time::Duration`, but it would be better to keep the internal representation as u64 for implementation efficiency.